Fix the derivative of `acosh` for complex numbers.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80841

For a summary of what was going on, see https://github.com/pytorch/pytorch/issues/50692#issuecomment-1173922441

Fixes https://github.com/pytorch/pytorch/issues/50692

lezcano committed:
Fix the derivative of `acosh` for complex numbers.

For a summary of what was going on, see
https://github.com/pytorch/pytorch/issues/50692#issuecomment-1173922441

Fixes
https://github.com/pytorch/pytorch/issues/50692

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -277,7 +277,7 @@
 
 - name: acosh(Tensor self) -> Tensor
-  self: grad * (self.pow(2) - 1).rsqrt().conj()
+  self: "self.is_complex() ? grad * ((self + 1).rsqrt() * (self - 1).rsqrt()).conj() : grad * (self * self - 1).rsqrt()"
```
The two are mathematically the same at a glance? Why can't we keep only one?

Lezcano replied:
The first one is correct for both real and complex numbers. The second one is a simplification for real numbers.

albanD replied:
Can you leave a comment there: save one addition by re-ordering the terms for the real case.

Lezcano replied:
Ugh, of course I wrote "compelx"... lol

albanD replied:
This can be added to any other PR :D

albanD reviewed (approved):
I think a small comment to explain the duplication would be nice.
LGTM otherwise!

lezcano committed:
Update on "Fix the derivative of `acosh` for complex numbers."


For a summary of what was going on, see https://github.com/pytorch/pytorch/issues/50692#issuecomment-1173922441

Fixes https://github.com/pytorch/pytorch/issues/50692

[ghstack-poisoned]

anjali411 commented:
Similarly looks like atanh derivative is also wrong. It should be: `grad * 1 / ((1 - self) * (1 + self)).conj()`

Lezcano commented:
I think the derivative for the `atanh` is right.

Lezcano commented:
Re the edit. Those notes are for real numbers. The formula for the solution of a quadratic equation just works for real coefficients. For complex coefficients, you need to "complete the square" (which is the way to prove the usual formula for the quadratic equation really) and be careful when taking square roots and logarithms, as these are multivalued functions.

If you to simply get a formula, ask what's `asin'(z)` to wolfram alpha.

Lezcano commented:
fwiw, cppreference is also careful of defining things the right way in https://en.cppreference.com/w/cpp/numeric/complex/acosh.
